### PR TITLE
Adding Support for `json_populate_record` and `jsonb_populate_record`

### PR DESCRIPTION
**Pull request by prkbuilds**
Using the traits already defined and have written the test_cases with my understanding of the functions, but open to suggestions and learnings!

### CODE REVIEW

**Comment by PratikFandade** [comment]
just for context the PR is for the issue #4216

**Review by weiznich — CHANGES_REQUESTED** [review]
I'm sorry to take that long to review this PR and I'm also sorry to bring up an issue, but it seems like the nullable bounds are not correct yet. I've added a more detailed comment in the specific location on how this can be fixed.

File: diesel/src/pg/expression/functions.rs [hunk 1]
```diff
@@ -2279,3 +2279,121 @@ define_sql_function! {
...
+    /// assert_eq!(record, ("John".to_string(), 42));
+    ///
+    /// #    Ok(())
+    /// # }
+    /// ```
+    #[sql_name = "jsonb_populate_record"]
+    fn jsonb_populate_record<
+        B: RecordOrNullableRecord + SingleValue,
+        J: JsonbOrNullableJsonb + CombinedNullableValue<Jsonb, B>
+    >(base: B, from_json: J) -> J::Out;
```
[the remark this answers is not in the document]
Comment (PratikFandade) [reply diesel/src/pg/expression/functions.rs:2398]: Thanks @weiznich ! This is great to know and really informative, I totally understand what needs to be done, I'll implement the required change are add the changes. and thanks for telling a good way to test the nullability, will definitely keep that in mind next time.

**Comment by PratikFandade** [comment]
I'm not sure if the naming convention is the best for the new Nullable Trait implementation. I think this will take care of Nullablility, but let me know if there is anything I might have missed.

File: diesel/src/pg/expression/expression_methods.rs [hunk 2]
```diff
@@ -3726,4 +3726,19 @@ pub(in crate::pg) mod private {
 
     impl<T> RecordOrNullableRecord for Record<T> {}
     impl<T> RecordOrNullableRecord for Nullable<Record<T>> {}
+
+    pub trait CombinedOneNullableValue<O, Out>: SingleValue {
```
Comment (weiznich) [inline diesel/src/pg/expression/expression_methods.rs:3730]: Can we name this `CombinedAllNullableValue` or something like that instead to transport that this only evaluates to `Nullable` if all inputs are `Nullable`?

**Review by weiznich — COMMENTED** [review]
Thanks for the update. This looks good beside the name of the new trait.

File: diesel/src/pg/expression/expression_methods.rs (hunk 2 above)
Comment (PratikFandade) [reply diesel/src/pg/expression/expression_methods.rs:3730]: Hi @weiznich , sure I will change the name accordingly. And yes they allarenullable only returns nullable value when both values are null, completing the logic we need for this trait.

**Review by weiznich — APPROVED** [review]
Thanks for the update

**Pull request merged**